spacy_udpipe/udpipe.py

```diff
@@ -0,0 +1,202 @@
+        for i, (token, next_token) in enumerate(zip(tokens[:-1], tokens[1:])):
+            word = sentence.addWord(token)
+            # word.id = i
+            if re.match(r"\W", next_token):
+                # leave no space after current token iff next token
+                # is non-alphanumeric (i.e. punctuation)
+                word.misc = self.NO_SPACE
+        return True
+
+
+class UDPipeModel(object):
+
+    def __init__(
+        self,
+        lang: str,
+        path: Optional[str] = None,
+        meta: Optional[Dict] = None
+    ):
+        """Load UDPipe model for given language.
+
+        lang: ISO 639-1 language code or shorthand UDPipe model name.
+        path: Path to UDPipe model.
+        meta: Meta-information about the UDPipe model.
+        """
+        path = path or get_path(lang=lang)
+        self.model = Model.load(path)
+        if self.model is None:
+            raise Exception(f"Cannot load UDPipe model from file '{path}'")
+        self._lang = lang.split("-")[0]
+        self._meta = meta or {"author": "Milan Straka & Jana Straková",
+                              "description": "UDPipe pretrained model.",
+                              "email": "straka@ufal.mff.cuni.cz",
+                              "lang": f"udpipe_{self._lang}",
+                              "license": "CC BY-NC-SA 4.0",
+                              "name": path.split("/")[-1],
+                              "parent_package": "spacy_udpipe",
+                              "pipeline": [
+                                  "Tokenizer", "Tagger", "Lemmatizer", "Parser"
+                              ],
+                              "source": "Universal Dependencies 2.5",
+                              "url": "http://ufal.mff.cuni.cz/udpipe",
+                              "version": "1.2.0"
+                              }
+
+    def __call__(
+        self,
+        text: Union[
+            str,
+            List[str],
+            List[List[str]]
+        ]
+    ) -> List[Sentence]:
+        """Tokenize, tag and parse the text and return it in an UDPipe
+        representation.
+
+        text: Input text, can be presegmented or pretokenized:
+            str             : raw text,
+            List[str]       : presegmented text,
+            List[List[str]] : pretokenized text.
+        RETURNS: Processed sentences.
+        """
+        sentences = self.tokenize(text)
+        for s in sentences:
+            self.tag(s)
+            self.parse(s)
+        return sentences
+
+    def _read(self, text: str, input_format: str) -> List[Sentence]:
+        """Convert the text to an UDPipe representation.
+
+        text: Input text.
+        input_format: Desired input format.
+        RETURNS: Processed sentences.
+        """
+        input_format.setText(text)
+        error = ProcessingError()
+        sentences = []
+
+        sentence = Sentence()
+        while input_format.nextSentence(sentence, error):
```

Comment:
perhaps a cleaner solution would be to yield the sentence in each iteration. That is much more in line what typical iterators do and then you can simply have a for loop instead.

Reply:
A valid point, however, these lines actually call [C++ code via Python bindings](https://pypi.org/project/ufal.udpipe/).